File: csharp/Link.Foundation.Links.Notation.Tests/MultiQuoteParserTests.cs
using System;
using Xunit;

namespace Link.Foundation.Links.Notation.Tests
{
    public static class MultiQuoteParserTests
    {
        // Helper to extract single reference ID
        private static string? GetSingleRefId(System.Collections.Generic.IList<Link<string>> result)
        {
            if (result.Count == 1 && result[0].Id == null && result[0].Values?.Count == 1)
            {
                return result[0].Values[0].Id;
            }
            return result.Count == 1 ? result[0].Id : null;
        }

        // ============================================================================
        // Backtick Quote Tests (Single Backtick)
        // ============================================================================

        [Fact]
        public static void TestBacktickQuotedReference()
        {
            var parser = new Parser();
            var result = parser.Parse("`backtick quoted`");
            Assert.Equal("backtick quoted", GetSingleRefId(result));
        }

        [Fact]
        public static void TestBacktickQuotedWithSpaces()
        {
            var parser = new Parser();
            var result = parser.Parse("`text with spaces`");
            Assert.Equal("text with spaces", GetSingleRefId(result));
        }

        [Fact]
        public static void TestBacktickQuotedMultiline()
        {
            var parser = new Parser();
            var result = parser.Parse("(`line1\nline2`)");
            Assert.Single(result);
            Assert.NotNull(result[0].Values);
            Assert.Single(result[0].Values);
            Assert.Equal("line1\nline2", result[0].Values![0].Id);
        }

        [Fact]
        public static void TestBacktickQuotedWithEscapedBacktick()
        {
            var parser = new Parser();
            var result = parser.Parse("`text with `` escaped backtick`");
            Assert.Equal("text with ` escaped backtick", GetSingleRefId(result));
        }

        // ============================================================================
        // Single Quote Tests (with escaping)
        // ============================================================================

        [Fact]
        public static void TestSingleQuoteWithEscapedSingleQuote()
        {
            var parser = new Parser();
            var result = parser.Parse("'text with '' escaped quote'");
            Assert.Equal("text with ' escaped quote", GetSingleRefId(result));
        }

        // ============================================================================
        // Double Quote Tests (with escaping)
        // ============================================================================

        [Fact]
        public static void TestDoubleQuoteWithEscapedDoubleQuote()
        {
            var parser = new Parser();
            var result = parser.Parse("\"text with \"\" escaped quote\"");
            Assert.Equal("text with \" escaped quote", GetSingleRefId(result));
        }

        // ============================================================================
        // Double Quotes (2 quote chars) Tests
        // ============================================================================

        [Fact]
        public static void TestDoubleDoubleQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("\"\"double double quotes\"\"");
            Assert.Equal("double double quotes", GetSingleRefId(result));
        }

        [Fact]
        public static void TestDoubleDoubleQuotesWithSingleQuoteInside()
        {
            var parser = new Parser();
            var result = parser.Parse("\"\"text with \" inside\"\"");
            Assert.Equal("text with \" inside", GetSingleRefId(result));
        }

        [Fact]
        public static void TestDoubleDoubleQuotesWithEscape()
        {
            var parser = new Parser();
            var result = parser.Parse("\"\"text with \"\"\"\" escaped double\"\"");
            Assert.Equal("text with \"\" escaped double", GetSingleRefId(result));
        }

        [Fact]
        public static void TestDoubleSingleQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("''double single quotes''");
            Assert.Equal("double single quotes", GetSingleRefId(result));
        }

        [Fact]
        public static void TestDoubleSingleQuotesWithSingleQuoteInside()
        {
            var parser = new Parser();
            var result = parser.Parse("''text with ' inside''");
            Assert.Equal("text with ' inside", GetSingleRefId(result));
        }

        [Fact]
        public static void TestDoubleSingleQuotesWithEscape()
        {
            var parser = new Parser();
            var result = parser.Parse("''text with '''' escaped single''");
            Assert.Equal("text with '' escaped single", GetSingleRefId(result));
        }

        [Fact]
        public static void TestDoubleBacktickQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("``double backtick quotes``");
            Assert.Equal("double backtick quotes", GetSingleRefId(result));
        }

        [Fact]
        public static void TestDoubleBacktickQuotesWithBacktickInside()
        {
            var parser = new Parser();
            var result = parser.Parse("``text with ` inside``");
            Assert.Equal("text with ` inside", GetSingleRefId(result));
        }

        [Fact]
        public static void TestDoubleBacktickQuotesWithEscape()
        {
            var parser = new Parser();
            var result = parser.Parse("``text with ```` escaped backtick``");
            Assert.Equal("text with `` escaped backtick", GetSingleRefId(result));
        }

        // ============================================================================
        // Triple Quotes (3 quote chars) Tests
        // ============================================================================

        [Fact]
        public static void TestTripleDoubleQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("\"\"\"triple double quotes\"\"\"");
            Assert.Equal("triple double quotes", GetSingleRefId(result));
        }

        [Fact]
        public static void TestTripleDoubleQuotesWithDoubleQuoteInside()
        {
            var parser = new Parser();
            var result = parser.Parse("\"\"\"text with \"\" inside\"\"\"");
            Assert.Equal("text with \"\" inside", GetSingleRefId(result));
        }

        [Fact]
        public static void TestTripleDoubleQuotesWithEscape()
        {
            var parser = new Parser();
            var result = parser.Parse("\"\"\"text with \"\"\"\"\"\" escaped triple\"\"\"");
            Assert.Equal("text with \"\"\" escaped triple", GetSingleRefId(result));
        }

        [Fact]
        public static void TestTripleSingleQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("'''triple single quotes'''");
            Assert.Equal("triple single quotes", GetSingleRefId(result));
        }

        [Fact]
        public static void TestTripleBacktickQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("```triple backtick quotes```");
            Assert.Equal("triple backtick quotes", GetSingleRefId(result));
        }

        // ============================================================================
        // Quadruple Quotes (4 quote chars) Tests
        // ============================================================================

        [Fact]
        public static void TestQuadrupleDoubleQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("\"\"\"\"quadruple double quotes\"\"\"\"");
            Assert.Equal("quadruple double quotes", GetSingleRefId(result));
        }

        [Fact]
        public static void TestQuadrupleSingleQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("''''quadruple single quotes''''");
            Assert.Equal("quadruple single quotes", GetSingleRefId(result));
        }

        [Fact]
        public static void TestQuadrupleBacktickQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("````quadruple backtick quotes````");
            Assert.Equal("quadruple backtick quotes", GetSingleRefId(result));
        }

        // ============================================================================
        // Quintuple Quotes (5 quote chars) Tests
        // ============================================================================

        [Fact]
        public static void TestQuintupleDoubleQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("\"\"\"\"\"quintuple double quotes\"\"\"\"\"");
            Assert.Equal("quintuple double quotes", GetSingleRefId(result));
        }

        [Fact]
        public static void TestQuintupleSingleQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("'''''quintuple single quotes'''''");
            Assert.Equal("quintuple single quotes", GetSingleRefId(result));
        }

        [Fact]
        public static void TestQuintupleBacktickQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("`````quintuple backtick quotes`````");
            Assert.Equal("quintuple backtick quotes", GetSingleRefId(result));
        }

        // ============================================================================
        // Complex Scenarios Tests
        // ============================================================================

        [Fact]
        public static void TestMixedQuotesInLink()
        {
            var parser = new Parser();
            var result = parser.Parse("(\"double\" 'single' `backtick`)");
            Assert.Single(result);
            Assert.NotNull(result[0].Values);
            Assert.Equal(3, result[0].Values!.Count);
            Assert.Equal("double", result[0].Values[0].Id);
            Assert.Equal("single", result[0].Values[1].Id);
            Assert.Equal("backtick", result[0].Values[2].Id);
        }

        [Fact]
        public static void TestBacktickAsIdInLink()
        {
            var parser = new Parser();
            var result = parser.Parse("(`myId`: value1 value2)");
            Assert.Single(result);
            Assert.Equal("myId", result[0].Id);
            Assert.NotNull(result[0].Values);
            Assert.Equal(2, result[0].Values!.Count);
        }

        [Fact]
        public static void TestCodeBlockLikeContent()
        {
            var parser = new Parser();
            var result = parser.Parse("```const x = 1;```");
            Assert.Equal("const x = 1;", GetSingleRefId(result));
        }

        [Fact]
        public static void TestNestedQuotesInMarkdown()
        {
            var parser = new Parser();
            var result = parser.Parse("``Use `code` in markdown``");
            Assert.Equal("Use `code` in markdown", GetSingleRefId(result));
        }

        [Fact]
        public static void TestJsonStringWithQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("\"\"{ \"key\": \"value\"}\"\"");
            Assert.Equal("{ \"key\": \"value\"}", GetSingleRefId(result));
        }

        // ============================================================================
        // Edge Cases
        // ============================================================================

        [Fact]
        public static void TestWhitespacePreservedInQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("\"  spaces  \"");
            Assert.Equal("  spaces  ", GetSingleRefId(result));
        }

        [Fact]
        public static void TestMultilineInDoubleDoubleQuotes()
        {
            var parser = new Parser();
            var result = parser.Parse("(\"\"line1\nline2\"\")");
            Assert.Single(result);
            Assert.NotNull(result[0].Values);
            Assert.Single(result[0].Values);
            Assert.Equal("line1\nline2", result[0].Values![0].Id);
        }
    }
}

File: rust/tests/multi_quote_parser_tests.rs
use links_notation::{parse_lino, LiNo};

// Helper to extract the single reference from a parsed result
fn get_single_ref_id(lino: &LiNo<String>) -> Option<&String> {
    match lino {
        LiNo::Ref(id) => Some(id),
        LiNo::Link { id: None, values } if values.len() == 1 => {
            if let LiNo::Ref(id) = &values[0] {
                Some(id)
            } else if let LiNo::Link {
                id: Some(ref_id),
                values: inner_values,
            } = &values[0]
            {
                if inner_values.is_empty() {
                    Some(ref_id)
                } else {
                    None
                }
            } else {
                None
            }
        }
        LiNo::Link {
            id: Some(ref_id),
            values,
        } if values.is_empty() => Some(ref_id),
        _ => None,
    }
}

// Helper to get values from a link
fn get_values(lino: &LiNo<String>) -> Option<&Vec<LiNo<String>>> {
    match lino {
        LiNo::Link { values, .. } => {
            // If it's a wrapper link (outer link)
            if values.len() == 1 {
                if let LiNo::Link {
                    values: inner_values,
                    ..
                } = &values[0]
                {
                    return Some(inner_values);
                }
            }
            Some(values)
        }
        _ => None,
    }
}

// ============================================================================
// Backtick Quote Tests (Single Backtick)
// ============================================================================

#[test]
fn test_backtick_quoted_reference() {
    let result = parse_lino("`backtick quoted`").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"backtick quoted".to_string())
    );
}

#[test]
fn test_backtick_quoted_with_spaces() {
    let result = parse_lino("`text with spaces`").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with spaces".to_string())
    );
}

#[test]
fn test_backtick_quoted_multiline() {
    let result = parse_lino("(`line1\nline2`)").unwrap();
    if let LiNo::Link { values, .. } = &result {
        if let Some(inner) = values.first() {
            if let LiNo::Link {
                values: inner_vals, ..
            } = inner
            {
                if let Some(LiNo::Ref(id)) = inner_vals.first() {
                    assert_eq!(id, "line1\nline2");
                    return;
                }
            }
            if let LiNo::Ref(id) = inner {
                assert_eq!(id, "line1\nline2");
                return;
            }
        }
    }
    panic!("Expected multiline backtick content");
}

#[test]
fn test_backtick_quoted_with_escaped_backtick() {
    let result = parse_lino("`text with `` escaped backtick`").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with ` escaped backtick".to_string())
    );
}

// ============================================================================
// Single Quote Tests (with escaping)
// ============================================================================

#[test]
fn test_single_quote_with_escaped_single_quote() {
    let result = parse_lino("'text with '' escaped quote'").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with ' escaped quote".to_string())
    );
}

// ============================================================================
// Double Quote Tests (with escaping)
// ============================================================================

#[test]
fn test_double_quote_with_escaped_double_quote() {
    let result = parse_lino("\"text with \"\" escaped quote\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with \" escaped quote".to_string())
    );
}

// ============================================================================
// Double Quotes (2 quote chars) Tests
// ============================================================================

#[test]
fn test_double_double_quotes() {
    let result = parse_lino("\"\"double double quotes\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"double double quotes".to_string())
    );
}

#[test]
fn test_double_double_quotes_with_single_quote_inside() {
    let result = parse_lino("\"\"text with \" inside\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with \" inside".to_string())
    );
}

#[test]
fn test_double_double_quotes_with_escape() {
    let result = parse_lino("\"\"text with \"\"\"\" escaped double\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with \"\" escaped double".to_string())
    );
}

#[test]
fn test_double_single_quotes() {
    let result = parse_lino("''double single quotes''").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"double single quotes".to_string())
    );
}

#[test]
fn test_double_single_quotes_with_single_quote_inside() {
    let result = parse_lino("''text with ' inside''").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with ' inside".to_string())
    );
}

#[test]
fn test_double_single_quotes_with_escape() {
    let result = parse_lino("''text with '''' escaped single''").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with '' escaped single".to_string())
    );
}

#[test]
fn test_double_backtick_quotes() {
    let result = parse_lino("``double backtick quotes``").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"double backtick quotes".to_string())
    );
}

#[test]
fn test_double_backtick_quotes_with_backtick_inside() {
    let result = parse_lino("``text with ` inside``").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with ` inside".to_string())
    );
}

#[test]
fn test_double_backtick_quotes_with_escape() {
    let result = parse_lino("``text with ```` escaped backtick``").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with `` escaped backtick".to_string())
    );
}

// ============================================================================
// Triple Quotes (3 quote chars) Tests
// ============================================================================

#[test]
fn test_triple_double_quotes() {
    let result = parse_lino("\"\"\"triple double quotes\"\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"triple double quotes".to_string())
    );
}

#[test]
fn test_triple_double_quotes_with_double_quote_inside() {
    let result = parse_lino("\"\"\"text with \"\" inside\"\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with \"\" inside".to_string())
    );
}

#[test]
fn test_triple_double_quotes_with_escape() {
    let result = parse_lino("\"\"\"text with \"\"\"\"\"\" escaped triple\"\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with \"\"\" escaped triple".to_string())
    );
}

#[test]
fn test_triple_single_quotes() {
    let result = parse_lino("'''triple single quotes'''").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"triple single quotes".to_string())
    );
}

#[test]
fn test_triple_single_quotes_with_double_quote_inside() {
    let result = parse_lino("'''text with '' inside'''").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with '' inside".to_string())
    );
}

#[test]
fn test_triple_single_quotes_with_escape() {
    let result = parse_lino("'''text with '''''' escaped triple'''").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with ''' escaped triple".to_string())
    );
}

#[test]
fn test_triple_backtick_quotes() {
    let result = parse_lino("```triple backtick quotes```").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"triple backtick quotes".to_string())
    );
}

#[test]
fn test_triple_backtick_quotes_with_double_backtick_inside() {
    let result = parse_lino("```text with `` inside```").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with `` inside".to_string())
    );
}

#[test]
fn test_triple_backtick_quotes_with_escape() {
    let result = parse_lino("```text with `````` escaped triple```").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with ``` escaped triple".to_string())
    );
}

// ============================================================================
// Quadruple Quotes (4 quote chars) Tests
// ============================================================================

#[test]
fn test_quadruple_double_quotes() {
    let result = parse_lino("\"\"\"\"quadruple double quotes\"\"\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"quadruple double quotes".to_string())
    );
}

#[test]
fn test_quadruple_double_quotes_with_triple_quote_inside() {
    let result = parse_lino("\"\"\"\"text with \"\"\" inside\"\"\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with \"\"\" inside".to_string())
    );
}

#[test]
fn test_quadruple_single_quotes() {
    let result = parse_lino("''''quadruple single quotes''''").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"quadruple single quotes".to_string())
    );
}

#[test]
fn test_quadruple_backtick_quotes() {
    let result = parse_lino("````quadruple backtick quotes````").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"quadruple backtick quotes".to_string())
    );
}

// ============================================================================
// Quintuple Quotes (5 quote chars) Tests
// ============================================================================

#[test]
fn test_quintuple_double_quotes() {
    let result = parse_lino("\"\"\"\"\"quintuple double quotes\"\"\"\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"quintuple double quotes".to_string())
    );
}

#[test]
fn test_quintuple_double_quotes_with_quad_quote_inside() {
    let result = parse_lino("\"\"\"\"\"text with \"\"\"\" inside\"\"\"\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"text with \"\"\"\" inside".to_string())
    );
}

#[test]
fn test_quintuple_single_quotes() {
    let result = parse_lino("'''''quintuple single quotes'''''").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"quintuple single quotes".to_string())
    );
}

#[test]
fn test_quintuple_backtick_quotes() {
    let result = parse_lino("`````quintuple backtick quotes`````").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"quintuple backtick quotes".to_string())
    );
}

// ============================================================================
// Complex Scenarios Tests
// ============================================================================

#[test]
fn test_mixed_quotes_in_link() {
    let result = parse_lino("(\"double\" 'single' `backtick`)").unwrap();
    if let Some(values) = get_values(&result) {
        assert_eq!(values.len(), 3);
        if let LiNo::Ref(id) = &values[0] {
            assert_eq!(id, "double");
        }
        if let LiNo::Ref(id) = &values[1] {
            assert_eq!(id, "single");
        }
        if let LiNo::Ref(id) = &values[2] {
            assert_eq!(id, "backtick");
        }
    } else {
        panic!("Expected values in link");
    }
}

#[test]
fn test_backtick_as_id_in_link() {
    let result = parse_lino("(`myId`: value1 value2)").unwrap();
    if let LiNo::Link { values, .. } = &result {
        if let Some(LiNo::Link {
            id,
            values: inner_values,
        }) = values.first()
        {
            assert_eq!(id.as_deref(), Some("myId"));
            assert_eq!(inner_values.len(), 2);
            return;
        }
    }
    panic!("Expected link with backtick id");
}

#[test]
fn test_code_block_like_content() {
    let result = parse_lino("```const x = 1;```").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"const x = 1;".to_string())
    );
}

#[test]
fn test_nested_quotes_in_markdown() {
    let result = parse_lino("``Use `code` in markdown``").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"Use `code` in markdown".to_string())
    );
}

#[test]
fn test_json_string_with_quotes() {
    let result = parse_lino("\"\"{ \"key\": \"value\"}\"\"").unwrap();
    assert_eq!(
        get_single_ref_id(&result),
        Some(&"{ \"key\": \"value\"}".to_string())
    );
}

// ============================================================================
// Edge Cases
// ============================================================================

#[test]
fn test_whitespace_preserved_in_quotes() {
    let result = parse_lino("\"  spaces  \"").unwrap();
    assert_eq!(get_single_ref_id(&result), Some(&"  spaces  ".to_string()));
}

#[test]
fn test_multiline_in_double_double_quotes() {
    let result = parse_lino("(\"\"line1\nline2\"\")").unwrap();
    if let Some(values) = get_values(&result) {
        if let Some(LiNo::Ref(id)) = values.first() {
            assert_eq!(id, "line1\nline2");
            return;
        }
    }
    panic!("Expected multiline content in double double quotes");
}
